libgraph/include/katana/GraphTopology.h

```diff
@@ -172,8 +178,47 @@ class KATANA_EXPORT GraphTopology : public GraphTopologyTypes {
     return edge_property_index(eid);
   }
 
+  // We promote these methods to the base topology to make it easier to search among all
+  // already cached topologies using the same predicate.
+  bool is_transposed() const noexcept {
+    return has_transpose_state(tsuba::RDGTopology::TransposeKind::kYes);
+  }
+
+  bool has_transpose_state(
+      const tsuba::RDGTopology::TransposeKind& expected) const noexcept {
+    return tpose_state_ == expected;
+  }
+
+  tsuba::RDGTopology::TransposeKind transpose_state() const noexcept {
+    return tpose_state_;
+  }
+
+  tsuba::RDGTopology::EdgeSortKind edge_sort_state() const noexcept {
+    return edge_sort_state_;
+  }
+
   void Print() const noexcept;
 
+protected:
+  void SetTransposeState(
```

Comment:
It is to enable derived classes to set the private members without giving them protected status, which is generally considered bad practice. I think they are only set in the constructors though, so it prob makes more sense to just have a GraphTopology constructor that takes them as input.

Reply:
@rob-isaac Even `protected`, you can't put them in the class initializer list... To your point, I specifically wanted to avoid these params being passed to the GraphTopology ctor, because for GraphTopology proper their values should be defaults. It's ugly, because we are trying to avoid virtual methods, otherwise I could have replaced member variables with virtual methods that return fixed values in GraphTopology proper.

Reply:
Why are we trying to avoid virtual members? Is there a reason to avoid them in all cases? Or is it reasonable to have them in specific cases where they are objectively cleaner than other approaches? 

Personally I would go with the virtual methods in this case and then benchmark it. Branch predictors are really good these days, so virtual call overhead can be quite low as long as each generated call instruction almost always calls the same target implementation. And I suspect that will be the case here since most parts of the system will always be using the same view.

Reply:
@mindv0rtex couldn't you maintain the current ctor, which sets the values to the defaults, but also maintain another ctor which explicitly sets these values? E.g:
```
public:
GraphTopology(<current_args>)
protected:
GraphTopology(<current_args>, shuffle_type, ect) // this will set the values
```
Then derived class constructors can call the second ctor to set the correct values instead of the defaults.
```
ShuffleTopology(<args>) : GraphTopology(<args>, /*shuffle type*/ true, ect), <other stuff>